server/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -368,6 +372,45 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
         builder.endObject();
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        QueryBuilder rewritten = super.doRewrite(queryRewriteContext);
+
+        if (rewritten == this && queryRewriteContext.getClass() == QueryRewriteContext.class) {
```

Comment:
I think we can refactor to avoid the `getClass()` comparison. This was the quick & dirty approach.

And just to clarify, this rewrite phase does not run in the can_match phase. It runs just before the can_match phase. It may seem like a pedantic detail, but it's important because the can_match phase is skipped for most search requests.

Reply:
> And just to clarify, this rewrite phase does not run in the can_match phase. It runs just before the can_match phase. It may seem like a pedantic detail, but it's important because the can_match phase is skipped for most search requests.


Understood - my comment referred to the current `CoordinatorRewriteContext`, which is really what runs on the can_match phase instead of actually being what is used in the coordinator.